### ISSUE TITLE
Add Timestamped Nodes support to Send Operation at Client side

### ISSUE THREAD

**Issue by sbernard31** [eclipse/leshan#1228]
Following #1192 we decide to add Timestamped nodes to Send Operation.

The Server side part was done by #1218.

This issue aims to track / discuss about client side support.

(Here some idea about a `DataCollector` ? => https://github.com/eclipse/leshan/issues/1192#issuecomment-1022220554)

**Comment by adamsero** [comment]
Hi, my name is Adam and I replaced @Michal-Wadowski as a contributor from Orange Polska. I've been studying the code and I made some first steps and committed them on [opl/timestamp_nodes_support_for_client](https://github.com/Przem83/leshan/tree/opl/timestamp_nodes_support_for_client).

I created a basic encoder for type TimestampedLwM2mNodes, with optional support for non-timestamped multiple nodes (in that case it uses the old encoder). So far I only tested it in a crude way, but I'm currently working on proper integration tests.

As for now, it is possible to send nodes with timestamps, but it's not possible to retrieve them from enablers - I'd like to discuss how to approach this topic after you approve of the encoder/tests.

**Comment by sbernard31** [comment]
(I will look at this tomorrow)

**Comment by sbernard31** [comment]
Hi Adam, welcome on board :wink: 

I looked quickly at the code and that sounds  to go in the right direction.

> So far I only tested it in a crude way, but I'm currently working on proper integration tests.

I'm not sure you will be able to start by integration test. You will probably need to define the client API before you write integration tests ? 

But you can start to add encoders unit tests. 

> As for now, it is possible to send nodes with timestamps, but it's not possible to retrieve them from enablers - I'd like to discuss how to approach this topic after you approve of the encoder/tests.

About client side API, I'm not sure it makes sense that enablers returns timestamped value. 
I rather have the idea of a kind of DataCollector (https://github.com/eclipse/leshan/issues/1192#issuecomment-1022220554)

The idea is to have an object which is responsible to collect data from enablers, timestamped the data, store it, then we use this stored data to send `SendRequest` later.

e.g. I configure my DataCollector to collect `/3/0/9` (battery level) each 10 minutes. this will get data like with `LeshanClient.collectData()`, and timestamp it with current timestamp then store it in memory.  Then e.g. each 2 hours I can send the collected data.

This is not crystal clear to me how this should looks like exactly but I think this could be a good way to explore. What do you think ? 

Anyway, If you really need timestamp from enablers, please let me know the use case :slightly_smiling_face:

**Comment by sbernard31** [comment]
(By the way I will be unavailable from the end of day to Monday 18th april, back on Tuesday)

**Comment by adamsero** [comment]
Alright, I will work on unit tests for encoders then - you're right about that.

> About client side API, I'm not sure it makes sense that enablers returns timestamped value.
> I rather have the idea of a kind of DataCollector
I agree with that, it's better not to modify the existing structure of enablers, but rather add new functionalities on top of the existing ones. However building a DataCollector is going to be a lot of work.

**Comment by sbernard31** [comment]
> However writing a DataCollector is going to be a lot of work.

I'm not so sure. :thinking: 

I guess we should define an interface with a very simple default implementation in a first time.
Then user will be able to provide their own implementation or we will be able to enhance it. 
Maybe in a first time we don't even need to provide an automatic periodically way to collect. (just a manual way is enough?)

Do not hesitate to ask for help if needed.

**Comment by adamsero** [comment]
I will think about how to implement this idea while you're away and hopefully have something ready by the time you're back 🤞. 

As for now, I added some unit tests for encoding in SenML-JSON and SenML-CBOR (f6ca130ad1b5df6fd5f15c115b60bf9e8e52276f)

**Comment by adamsero** [comment]
I think I should have something to show later today when it comes to Data Collector. In the meantime, do you possibly know if the next release is going to contain changes from #1218? And if so, when will it come out? It's fairly important for Orange to have these changes in a release sometime soon.

**Comment by sbernard31** [comment]
> In the meantime, do you possibly know if the next release is going to contain changes from https://github.com/eclipse/leshan/pull/1218? 

Yes, next release will integrate this. 

> And if so, when will it come out? 

The plan I had in mind was  : https://github.com/eclipse/leshan/issues/1222.
=> Waiting for `OSCORE` and `timestamped nodes for send` to release the **2.0.0-M7**.

> It's fairly important for Orange to have these changes in a release sometime soon.

What could be a good release date for you ? 

(Note that I will be often unavailable during May)

**Comment by sbernard31** [comment]
About `DataCollector`, I looked quickly at your code. You get the idea but I think we need to take time to specify this more precisely. I will try to take to time on this. 

Waiting and as first step you can create a PR with encoder modification + unit test about it (without data collector)? This should be an easy PR and so probably a first good step to enter in Leshan project ?

>    Add a separate interactive command that starts/stops periodic read on a specified resource (it could also create the collector, although that can be done statically in code)

Yes I also have this kind of idea about adding some command about this in leshan-client-demo.

> These collectors would be mapped by the resource path, so Map<LwM2mPath, DataCollector>

For now I don't know if we should have several DataCollector or only 1.

>  Add an option to send command that causes data to be read from the collector instead of the enabler. Collector would be retrieved by the path provided by the user

Same here I don't know if the collector is responsible to send or not the data.

I really need to think more about all of this.  :thinking:

**Comment by adamsero** [comment]
I created a PR #1246

**Comment by adamsero** [comment]
> What could be a good release date for you ?

It would be convenient for us if you could release a version with timestamped Send before OSCORE, especially since you mentioned you might be less active in May.

**Comment by sbernard31** [comment]
> It would be convenient for us if you could release a version with timestamped Send before OSCORE

You mean the server part only ? or also the client part which is currently in development ? 

Do you have an approximate date in mind ?  end of april ? mid May ? end of may ?

**Comment by sbernard31** [comment]
@adamsero let me know that as soon as possible (especially if you need it for end of April, because there is not so much time remaining :sweat_smile: )

Did you try to integrate the current `master` branch to be sure it behave like you expect ?

**Comment by sbernard31** [comment]
The next release will be mainly for you and I'm asking myself if we should add the last californium release [3.5.0](https://github.com/eclipse/californium/issues/2001) to this release  ? Currently we depends on cf 3.4.0.

**Comment by adamsero** [comment]
Sorry for the delay, but I had to contact team members and we established that the latest date would be around mid-May. It will be enough if only Server-side changes are in the release, but I don't know if we can make it in time to also include Client-side. We could have PR #1246 in the release at the very least (if you even want to include it).

**Comment by sbernard31** [comment]
OK so I think we can integrate #1246 and #1247.
For mid-May, I bet we will not be able to include the DataCollector in time. 

I will be unavailable from April 29th to May 8th (included),  So either I try to release before the end of the week or when I will be back the 9th. 

(Note I will be unavailable again from May 13th to 22th and also 26/27th, then I will be avaialble again as usual)

**Comment by sbernard31** [comment]
OK so let's focus on #1246 and #1247, then I try to make a release.

**Comment by sbernard31** [comment]
Tomorrow is my last day of work for April, so I need to integrate #1246 and #1247 in the morning to be able to release in time.
So if you want to do some modification you have until tomorrow morning. :grin: 
If you want to stop me :sweat_smile: ,  let me know.
If you haven't time to polish regarding to my comments on those PRs, not a big deal I will do the little cleaning myself :wink:.

**Comment by sbernard31** [comment]
(spoiler alert, this https://github.com/eclipse/leshan/issues/1224 could prevent me to release tomorrow)

**Comment by adamsero** [comment]
I made changes to both branches, should be good now.

**Comment by sbernard31** [comment]
> (spoiler alert, this https://github.com/eclipse/leshan/issues/1224 could prevent me to release tomorrow)

Lucky me, today the github quota was not reached , so release is ongoing. :relieved:

**Comment by sbernard31** [comment]
@adamsero The release 2.0.0-M7 is out : https://github.com/eclipse/leshan/issues/563#issuecomment-1112361967

I guess next time we should create a dedicated issue to talk about this kind of thing (e.g. new release) because we polluted a little the Timestamped node thread here :sweat_smile:.

**Comment by sbernard31** [comment]
@adamsero, sorry to not have finding time to give you more hint about https://github.com/eclipse/leshan/issues/1228#issuecomment-1106385910

I let you move forward on this alone for the next week (back 9th May)

**Comment by adamsero** [comment]
> @adamsero The release 2.0.0-M7 is out : [#563 (comment)](https://github.com/eclipse/leshan/issues/563#issuecomment-1112361967)
> 
> I guess next time we should create a dedicated issue to talk about this kind of thing (e.g. new release) because we polluted a little the Timestamped node thread here 😅.

True, but I didn't expect the release so quickly, originally I just wanted to ask about its availability. Once again, thanks for getting it done so fast 👍

**Comment by adamsero** [comment]
> I let you move forward on this alone for the next week (back 9th May)

Does that mean I should create a PoC of the ideas I mentioned? You seemed to be quite ambivalent about some of them:
>For now I don't know if we should have several DataCollector or only 1.

> Same here I don't know if the collector is responsible to send or not the data.

I just don't really want to code it when you haven't accepted the idea yet as it might be a waste of my and your time.

**Comment by adamsero** [comment]
@sbernard31 I worked ab it on management and scalability of Data Collectors and this is my proposition:
- `DefaultDataCollector` class is the default implementation of `DataCollector` interface and it provides basic functionality like starting and stopping periodic read and retrieving the collected data at any given moment (previously we had `TemperatureDataCollector` as a sample class in client demo but I decided to generalize it and make it more expandable so I renamed it and moved it to client core).
- `DataCollectorManager` class is used to store Data Collectors and makes it easy to retrieve collected data from multiple collectors at once by when provided a list of paths.
- Data Collectors are identified by the node that they are collecting data from and are accessible by providing appropriate path. `DataCollectorManager` contains a `LwM2mRootEnabler` object so collectors can easily read data from enablers and the manager acts as proxy.
- User can choose to use data from the collector instead of the current data from the enabler by choosing appropriate option in CLI when executing send (`-dc` or `--data-collector` flag). User can also choose to keep the stored data in the collector after retrieving it by adding `-kv` or `--keep-values` flag (by default the data is cleared). If user chooses to use `-dc`, an appropriate collector will be found using the path that user provided and the data collected from it will be sent. Providing a path to a node that doesn't exist or doesn't have a collector assigned to it will result in an error printed to CLI.

I implemented these ideas on [opl/timestamp_nodes_support_for_client](https://github.com/przem83/leshan/tree/opl/timestamp_nodes_support_for_client) branch in commits 12b6adbb7a392ac96bc5af3f522ca5e1ed86a427 and 3a9f24db269dabdee3d9d233c52a3f68e4c2f9c5. Let me know what you think about this solution or if you have any questions 🙂.

**Comment by sbernard31** [comment]
@adamsero 

I started to look at this. 
(:warning: as we are in a design phase my opinion can often change :sweat_smile: )

I think creating a `DefaultDataCollector` is better.

For now, I'm not sure a `DataCollector` should be limited to 1 path. :thinking:
And if it is not limited to 1 path then maybe we don't need a `DataCollectorManager`.
I think most of the time only 1 implementation will be used but if you need to use several `DataCollector` you can imagine a kind of `MultipleDataCollector` which implement `DataCollector` and delegate it works to inner `DataCollector`.

> DataCollectorManager contains a LwM2mRootEnabler object so collectors can easily read data from enablers and the manager acts as proxy.

This could be a good point about DataCollectorManager :thinking: 

About flush value, I think the use case is that you want to flush data only once you are sure sever receive the data :thinking: 
(but let's this aside for now)

:pray:  Thx for taking time to explore some solution. 

I would like to experiment some idea by writting some code. 
I will probably base my work on yours, could you eventually push your code in a PR ? (Don't clean it, it's just for exploration)

**Comment by sbernard31** [comment]
@adamsero I create a PR to share some idea : https://github.com/eclipse/leshan/pull/1255 (take a look at the comment in the code too, I tried to add some concerns I have)

Sorry for the state of this PR, I did something quick&dirty before to be absent again for 1 week.  So you have some material to think / experiment a bit more about this if you want. 

I feel that I block you a little on this feature and I hope this is not too frustrating for you. Sorry about that :pray: . After May, I think I will be able to provide you more prompt support.

**Comment by sbernard31** [comment]
I push some work about switch from `DataCollector` design to `DataSender` design. (See #1255 )
(I removed the DataCollector concept for now but if we want to share some behavior between `DataSender` we can reintroduce it)

**Comment by adamsero** [comment]
I think it's a good concept to work on. If you want me to, I can polish the code a little and get it ready for production. Or maybe you have some other changes that you want to introduce?

**Comment by sbernard31** [comment]
> Or maybe you have some other changes that you want to introduce?

Nothing more to add on my side. That was mainly to share the whole ideas.

> I think it's a good concept to work on. 

Good to hear :slightly_smiling_face: 

>  I can polish the code a little and get it ready for production. 

Yep, that sounds great.
I think a good first step could be to : 
- add the `ManualDataSender` + unit test
- add manual sender support to demo (eventually in a seperated PR and/or commit)

Then once we integrated this, we can eventually work on a `PeriodicalDataSender` and its integration in leshan-client-demo. 
(If you think this is something we must add just after, personnaly I have no strong opinion about this.)

**Comment by adamsero** [comment]
I was thinking if it would be possible to add [Mockito](https://site.mockito.org/) as a dependency - it would really make testing a lot easier (the scope of the dependency would only be `test`). If it's not possible, I don't really know how to test `ManualDataSender` as I don't know how to mock a root enabler otherwise.

**Comment by sbernard31** [comment]
By the past, we was using it and generally this doesn't make so great code.

In your case, you should be able to test ManualDataSender with a kind of fake `DataSenderManager`

But I guess maybe the simple way is just to write integration tests only : 
[SendTimestampedTest](https://github.com/eclipse/leshan/blob/master/leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/send/SendTimestampedTest.java) (You should remove the fake encoder and just use the new client API ?)

Note that this class will change with #1262  so maybe, I should integrate it before and you base your tests on it ?

**Comment by adamsero** [comment]
To be fair I decided not to include a `DataSenderManager` as we only have one `ManualDataSender` anyway (in the future we will have a `PeriodicDataSender` as well) so we don't really need such a class. My idea of a manager would be useful if we were to have multiple senders, each corresponding to a single enabler. 

That being said, my `ManualDataSender` class has a `LwM2mRootEnabler` directly within, so I guess I will try to mock it with some `FakeRootEnabler` kind of class. I actually want to write some unit tests for that as I think it would be useful - we also don't seem to have many unit tests in the project as a whole anyway.

**Comment by sbernard31** [comment]
> To be fair I decided not to include a DataSenderManager as we only have one ManualDataSender anyway (in the future we will have a PeriodicDataSender as well) so we don't really need such a class. My idea of a manager would be useful if we were to have multiple senders, each corresponding to a single enabler.

With my last design, I had in mind that maybe one day we will have both ManualDataSender and PeriodicDataSender used in leshan-client-demo. (each of them with their own collected data)

About DataSendManager, I was not totally satisfy about it but I feel that providing  `getCurrentValue` and `SendData` method could be useful for any DataSender.

But let's see how this will looks like without it :)

> That being said, my ManualDataSender class has a LwM2mRootEnabler directly within, so I guess I will try to mock it with some FakeRootEnabler kind of class. I actually want to write some unit tests for that as I think it would be useful -

I'm not 100% sure but I feel that  you make strong separation between unit tests and integration tests ?

My mistake when I was talking about unit-test  at  https://github.com/eclipse/leshan/issues/1228#issuecomment-1140884961, I should said *junit** test (I was talking about any automatic test execuable with Junit)

In this case, I feel that integration test could do the job better without need to create any mock, just by using the leshan API.
In a general way, I prefer the way which need to write less code.
I also like integration tests because they provides end-to-end tests. Generally this comes with execution times cost but in leshan we are pretty lucky because integration-tests are often quick to execute anyway. 

But if you feel unit tests is more appropriate go for it :)

>  we also don't seem to have many unit tests in the project as a whole anyway.

I guess more tests will be probably better (even if that often means more code to maintain too)
But just to be sure when you say unit tests, you exclude **leshan-integration-tests** or even with this one you feel there are not not so many tests ?

**Comment by sbernard31** [comment]
> I got around that by providing default methods to SendData interface so any class implementing it can re-use the methods without code redundancy. I

Yep, this is a solution to solve this. Sometime I forget we now use java8 as minimal version now :grin:   

I looked at your code and I feel there is maybe some drawback compare to previous design :
 1. less easy to add a custom DataSender.
 2. [sendData and getCurrentValue was reused by LwM2mClient API](https://github.com/eclipse/leshan/pull/1255/commits/720b2a33382fb8dce06c527a9ce6ecbc76d750a9#diff-514ddcf9bd574f88e0ceba6f6926c87fd102d484a5d5771609efe5c79fd93ecaR319-R340) that I feel not so bad.
 3. no way to use several implementation of DataSender. (I was not so sure about this idea at the beginning but after thinking more about it I feel it make sense now)
 
 What your opinion about those point ?

**Comment by adamsero** [comment]
> 1. less easy to add a custom DataSender.

That's true, I didn't think about passing `DataSender` object(s) to `LeshanClientBuilder`, instead one instance is always being created in `LeshanClient`'s constructor. We can easily change that by e.g. having `LeshanClientBuilder` accept a list of `DataSender` (though it makes little sense right now, more about that in 3.)

> 2. sendData and getCurrentValue was reused by LwM2mClient API that I feel not so bad.

I don't really understand what you mean

> 3. no way to use several implementation of DataSender. (I was not so sure about this idea at the beginning but after thinking more about it I feel it make sense now)

I don't see how a manager would help with different implementations, but maybe it could act as a "hub" for all of them. With your design, I don't really see the point of having more than one sender of each sub-type (manual, periodic etc.) though, especially since the manager's list of senders isn't exactly being used for anything important. User still has to have their senders in a separate container, since for collecting and sending data, you have to call a sender's method, which then refers to its manager, not the other way around.

In my opinion, a manager is only useful if we have multiple instances of data senders that can be created dynamically and/or the instances are somehow identifiable, so we can refer to them through the manager.

After all though, it's your call to make, I can adjust to whatever concept you feel like is the best - just give me your idea and I will go for it.

**Comment by sbernard31** [comment]
> I don't really understand what you mean

Sorry the link above (https://github.com/eclipse/leshan/issues/1228#issuecomment-1144653164) was not good, I fixed it (you can look at this again).

I meant that LwM2mClient.sendData() API was reusing API from Manager.

> I don't see how a manager would help with different implementations

One of its jobs could be to automatically start/stop/destroy DataSender, if needed. See [DataSenderManager](https://github.com/eclipse/leshan/blob/720b2a33382fb8dce06c527a9ce6ecbc76d750a9/leshan-client-core/src/main/java/org/eclipse/leshan/client/datacollector/DataSenderManager.java#L96-L123)

> but maybe it could act as a "hub" for all of them

I'm not sure what you mean by "hub", but I also ask myself if this is not just a kind of naming issue ? Maybe Manager is not the good word... :thinking: 

>  With your design, I don't really see the point of having more than one sender of each sub-type (manual, periodic etc.) 

We can not imagine which kind of sender user will implement :relaxed:, maybe they will need several instance of same sender.
But even with only manual and periodic I can imagine case where you want more than 1 of each sub-type.

Take your example above : _"e.g. a sensor wants to send temperature measurements every 30 minutes, but humidity measurements every 60 minutes."_
I understand 1 periodic sender for  temperature and 1 for humidity.

> In my opinion, a manager is only useful if we have multiple instances of data senders that can be created dynamically and/or the instances are somehow identifiable, so we can refer to them through the manager.

I also thought that adding sender dynamically could maybe be useful. But I don't know if we want to go that far ?
I can imagine some use cases, e.g. for **leshan-client-demo**, you can imagine that you when you call collect you give a name like : 
```
collect -in pool1 /6 // pool1 is the name of 1 instance of manual sender
collect -in pool1 /6
collect -in pool2 /3/0/1 // pool2 is the name of 1 instance of manual sender
send -from pool1 // which will send only /6 payload stored 
```
It's just an idea I don't know if this is the right move. 

(sorry to not give you more straightforward directive :grimacing:, I'm back on Tuesday)

**Comment by adamsero** [comment]
> I meant that LwM2mClient.sendData() API was reusing API from Manager.

Yeah, I agree, but "reusing API" doesn't really do much in this case, since instead of calling a `CaliforniumLwM2mRequestSender` object directly, we call `DataSenderManager`'s method, which is just one line of code and does the exact same thing. I can do it your way if you want it though, that's not a problem.

If we have multiple senders, we have to somehow identify them in order to send data only from selected senders, not from all of them. I Imagine that for periodic senders it doesn't matter, since they would send data without user's interaction (unless we want to adjust their settings dynamically) but for manual senders we need some identifying mechanism.

That's why I originally put a `Map<LwM2MPath, DataCollector>` in my `DataColletorManager` concept. While I agree that it's not ideal to map them by a path, giving them names or IDs makes more sense - that way we're not limited to once sender per node, but we can still get to them through CLI as a user.

I'm open to the manager idea, but IMO it would have to be responsible for more, i.e. holding the senders inside a map or something, since right now they are (or it is - a single one) stored awkwardly in `InteractiveCommands` as a class field. I can try and include the concept that you proposed with:
```
collect -in pool1 /6 // pool1 is the name of 1 instance of manual sender
collect -in pool1 /6
collect -in pool2 /3/0/1 // pool2 is the name of 1 instance of manual sender
send -from pool1 // which will send data stored in manual sender named "pool1"
```
What do you think?

**Comment by sbernard31** [comment]
(let me know if you prefer that I decide)

**Comment by adamsero** [comment]
> - without possibility to add/remove sender dynamically (this feature means handle start/stop dynamically too)
> - without allowing several pool in leshan-client-demo. (I mean we can keep 1 manual sender for demo for now if needed)

I'm assuming you meant **with** possibility, right? For now, we don't need to have dynamic creation/deletion (we can add that later) and as for starting/stopping, we'll get to that when we tackle periodic senders. 

Right now, I'd like to have a manager which holds named senders (names will be hard-coded for now). Every operation (collect/send) will go through the manager, which then will tell specified sender to collect data and hold it in its memory or retrieve collected data and send it. I think this way, we'll fully utilize the manager's potential, centralizing everything related to senders and clearly dividing the responsibilities - manager acts as proxy between the senders and user/californium client, while individual senders are responsible for more low-level tasks such as data collection and storage.

This is the concept I'd like to go with, I think it makes sense 🙂.

**Comment by sbernard31** [comment]
> I'm assuming you meant with possibility, right? For now, we don't need to have dynamic creation/deletion (we can add that later) and as for starting/stopping, we'll get to that when we tackle periodic senders.

Sorry I was not clear.
As my previous example with `sender identified by name` was about adding sender dynamically (pool1, pool2 ...).
I wanted to insist that you can go with `sender identified by name` but you don't have to implement dynamic stuff too. (unless you wanted it now)

I'm not sure it's more clear :sweat_smile: 

> This is the concept I'd like to go with, I think it makes sense slightly_smiling_face.

So let's go for this :+1:

**Comment by adamsero** [comment]
> As my previous example with `sender identified by name` was about adding sender dynamically (pool1, pool2 ...).

Oh, so in that example you could create pools at the same time as you collected data, got it. I was thinking more like:
1. User creates a sender with a command, giving the newly created sender a unique name.
2. When collecting/sending, user specifies to/from which sender and they have to enter a name of a sender that already exists. I guess they could also enter a new name along with an option like -create or something.

But we'll think about that another time, for now I'll implement what I described previously.

**Comment by adamsero** [comment]
I pushed the changes in 0647aad369e23f74ed9a7f100708fd011ecce9b9

**Comment by sbernard31** [comment]
I looked at this quickly and I don't get why  `DataSender`  interface has those methods ? :thinking: 
```java
void collectData(List<LwM2mPath> paths);

void sendCollectedData(ServerIdentity server, ContentFormat format, long timeoutInMs, boolean noFlush);
```

**Comment by adamsero** [comment]
Those methods are being called from `DataSenderManager`, and since it uses `Map<String, DataSender>` to hold the senders, I figured `DataSender` interface had to contain the definitions. I guess we could theoretically only have them in `ManualDataSender` and then downcast every instance of that class so we can call these methods - but the periodic data sender will probably also have the same methods in common, so I just put them in the interface.

**Comment by sbernard31** [comment]
Oops, sorry I didn't understand it like this :grimacing: 

I'm not sure I get how it will works with something else than manual sender ?

**Comment by sbernard31** [comment]
>  Since automatic senders (including periodic ones) don't require user interaction by definition, you only really need to provide access to the manual ones through the manager.

I'm not sure I agree :thinking:   
E.g.  if you want to change configuration dynamically you will need to access to automatic sender API. I can imagine a use case where you create a LWM2M object which control DataSender and so API can not be abstract. 

Or even leshan-client-demo do configure collect and send period.

> Of course, users don't have to use that option if they don't want to, but since the manager already aggregates senders, I thought it would be a good idea to have some utility methods like collectDataManually or sendDataManually - they just help with managing multiple manual senders.

In my opinion this add more code for not so much benefits because : 
 - We will not really be able  to do a whole abstract of manual sender. (We can imagine there is more than just collect/send maybe listCollected or persistcollected)
 - I feel that maybe abstracting those 2 methods are just too  **leshan-client-demo** oriented.
 
So I would prefer to keep it simple and so remove the methods in manger, but If you really want to added, let's go for it.

**Comment by adamsero** [comment]
OK, I can remove the `collectDataManually` and `sendDataManually` methods, but I believe I should keep 
`DataSender getDataSender(String senderName)`
and 
`<T extends DataSender> T getDataSender(String senderName, Class<T> senderSubType)`
since that will make it available for users to use the manager as a container for senders as well. Of course they don't have to hold them there, but it would be a nice feature IMO.

I'm not sure where to put the logic from the removed methods though, should it be directly in the `run()` in command's class?

**Comment by sbernard31** [comment]
> OK, I can remove the collectDataManually and sendDataManually methods, but I believe I should keep
`DataSender getDataSender(String senderName)`
and
`<T extends DataSender> T getDataSender(String senderName, Class<T> senderSubType)`

That sounds good.

> I'm not sure where to put the logic from the removed methods though, should it be directly in the run() in command's class?

Yep I think it makes sense too.

**Comment by adamsero** [comment]
I think it might be ready to integrate into master, unless you want it more polished?

**Comment by sbernard31** [comment]
Maybe I missed something but do you create a PR ? (or push --force on an existing one ?)

**Comment by adamsero** [comment]
No worries, I just wanted to wait for you to confirm it's finally ready. Unfortunately, there's going to be a merge conflict if I open a PR from [opl/manual_data_sender](https://github.com/przem83/leshan/tree/opl/manual_data_sender) due to the latest commit 30bad682481e97e956025a016b934d5264d0b36c on master. The forked repo maintainer from our company is on vacation, so I can't update our forked master to resolve the conflict. Any ideas what to do with that? He'll be back on Monday if anything.

**Comment by sbernard31** [comment]
> No worries, I just wanted to wait for you to confirm it's finally ready. 

I didn't look at this yet,  anyway PR is right place to discus about details in the code. 

> Unfortunately, there's going to be a merge conflict if I open a PR from [opl/manual_data_sender](https://github.com/przem83/leshan/tree/opl/manual_data_sender) due to the latest commit https://github.com/eclipse/leshan/commit/30bad682481e97e956025a016b934d5264d0b36c on master. The forked repo maintainer from our company is on vacation, so I can't update our forked master to resolve the conflict.

I'm not sure I get you ? you mean the `master` branch of [przem83/leshan](https://github.com/przem83/leshan) repo is 1 commit behind Leshan's `master`, correct ? 

But what is the problem exactly you don't know how to fetch this commit ? or you don't have the right to push on master in [przem83/leshan](https://github.com/przem83/leshan) ?

**Comment by sbernard31** [comment]
Ok so you can fetch the commit and rebase you work on it. 
When you will create the PR you will create it againts `master` from leshan repository (not przem83 one), so you don't care "your" master is 1 commit behind.

I hope I'm clear.

**Comment by adamsero** [comment]
By 'fetch it' do you mean open a PR from eclipse/master to przem83/manual_data_sender? If I do that, it shows me there will be conflicts:
![image](https://user-images.githubusercontent.com/37634022/173790379-f281f385-074e-4a01-b12b-5ce98b10a914.png)
If not, then I don't know what you mean by that. Sorry about my lack of knowledge, but I haven't worked with a public github repo and forks yet.

**Comment by sbernard31** [comment]
By 'fetch it', I mean git fetch this commit from leshan repository.
 
 I guess if you have conflict you didn't do it. (That's why I asked if _"you don't know how to fetch this commit "_ ?)
 I wanted to know if you wanted some help about it ?
 
In your local git repository you can have several remote repositories attached.  
Each repository have a name.   
Often beginner works with only 1 remote repository called `origin`.  
But you can add the official Leshan repository as remote too. (you can name it leshan)   
Then you will be able to fetch master from this repository.
Once it's done, juste rebase your work on commit 30bad6824, then push force your branch again. 

``` sh
# I suppose your current remote repo is named origin
# -------------------------------------------------------
# add leshan repo
git remote add leshan git@github.com:eclipse/leshan.git
# fetch leshan master branch
git fetch leshan master
# checkout your local branch if not already done 
git checkout manual_data_sender
# rebase it on leshan/master  (commit 30bad6824) 
git rebase leshan/master
# you will probably need to solve conflict
# once it's done, push force the branch 
# (if you are not so sure of you can create a local branch of previous origin/manual_data_sender branch)
git push origin  manual_data_sender --force
```
 
If I didn't missed anything,  you should be able to create a PR on leshan repo without conflicts.

**Comment by sbernard31** [comment]
(another solution just push the PR with the conflicts and I will resolve it when I will integrate it in `master` but I guess that maybe it could be useful to learn that kind of git stuff :point_up: )

**Comment by adamsero** [comment]
I created #1268

**Comment by adamsero** [comment]
I created #1268 but it didn't pass the code check - it looks like the compiler doesn't understand the `Consumer` interface in a `forEach` loop - since it's a Java 1.8 Interface, my question is can we actually use 1.8 or do I have to stick to 1.7?

**Comment by sbernard31** [comment]
I'm currently trying to add more automatic check (see https://github.com/eclipse/leshan/issues/1265)
Some comment should have been automatically added but it seems that there is some right issue with Leshan repository that I didn't face with my own repository (sbernard31/leshan)
I will try to investigate this.

About your question, this issue(https://github.com/eclipse/leshan/issues/924) should give you some answer.

I see that your PR contains merge commit. For Leshan we don't use it,  **rebase only**. 
I prepare a code check about this : https://github.com/eclipse/leshan/pull/1267
(By the no need to recreate new PR when this is about the same changes)

If you have more question relative to your PR, I think it's better to ask at #1268.

**Comment by sbernard31** [comment]
@adamsero, #1274 is now integrated in `master`, I think we can close this issue.

Do you want to work on Periodic Sender ?
From my point of view this is not a vital feature to have but if you want to work on this it's OK to me :slightly_smiling_face: 

In all case, I think we can close this issue and when needed create a new issue to talk about it ?

**Comment by adamsero** [comment]
Yeah, I think we can finally close this one now. I'm not sure about  Periodic Sender's priority right now considering #1222, but I might open it as a new issue at some point in the future. At the moment I have some side tasks but after I'm done with them we can think more about the issue.
